internal/provider/newrelic/provider_integration_test.go

```diff
@@ -123,12 +135,205 @@ func Test_Getting_external_metric_generates_a_query_not_rejected_by_backend(t *t
+	t.Run("when_returns_200_for_first_requests_and_then_500", func(t *testing.T) {
+		server := runHTTPTestServer(t, 3000, func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write([]byte(`{"data": {
+				"actor": {
+				  "account": {
+					"nrql": {
+					  "results": [{"one": 0.015}]
+					}
+				  }
+				}
+			}}`))
+		})
+
+		m := provider.ExternalMetricInfo{Metric: testMetricName}
+
+		// Make the 3rd request to fail.
+		for i := 0; i < 3; i++ {
+			if i == 2 {
+				server.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusInternalServerError)
+				})
+				if _, err := p.GetExternalMetric(ctx, "", nil, m); err == nil {
+					t.Error("Error expected")
+				}
+
+				return
+			}
+
+			if _, err := p.GetExternalMetric(ctx, "", nil, m); err != nil {
+				t.Errorf("Unexpected error = %v", err)
+			}
+		}
+
+		url := fmt.Sprintf("%s/apis/external.metrics.k8s.io/v1beta1/namespaces/*/test_metric", testEnv.BaseURL)
+		checkStatusCodeNotOK(ctx, t, &testEnv, url)
+	})
+}
+
+// checkStatusCodeNotOK sends a GET request to the given URL.
+//
+// If 401 or 403 return code is received, function will retry.
+//
+// If a status code = 2xx is received, function will fail the given test.
+func checkStatusCodeNotOK(ctx context.Context, t *testing.T, testEnv *testutil.TestEnv, url string) {
+	t.Helper()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		t.Fatalf("Creating request: %v", err)
+	}
+
+	if err := wait.PollImmediateUntilWithContext(testEnv.Context, 1*time.Second, func(context.Context) (bool, error) {
+		resp, err := testEnv.HTTPClient.Do(req)
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+				t.Fatalf("Test timed out: %v", err)
+			}
+
+			t.Logf("Fetching %s: %v", url, err)
+
+			time.Sleep(1 * time.Second)
+
+			return false, nil
+		}
+
+		defer func() {
+			_ = resp.Body.Close()
+		}()
+
+		switch resp.StatusCode {
+		// Generic API server does not wait for RequestHeaderAuthRequestController informers cache to be synchronized,
+		// so until this is done, metrics adapter will be responding with HTTP 403, so we want to retry on that.
+		case http.StatusForbidden, http.StatusUnauthorized:
+			t.Logf("Got %d response code, expected %d: %v. Retrying.", resp.StatusCode, http.StatusOK, resp)
+
+			return false, nil
+		case http.StatusOK:
+			t.Errorf("Got %d response code, expected != 2xx", resp.StatusCode)
+		default:
+		}
+
+		return true, nil
+	}); err != nil {
+		t.Fatalf("Timed out waiting for k8s apiserver to respond: %v", err)
+	}
+}
+
+func runHTTPTestServer(t *testing.T, port int, f func(w http.ResponseWriter, req *http.Request)) *httptest.Server {
+	t.Helper()
+
+	l, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
```

Comment:
Does the server _need_ to be in port 3000? HTTPtest is able to start on a random port and ask it which address it's listening on using the `.URL` property.

If we really need port to be 3000 it might be worth adding a comment explaining why.

Reply:
> Does the server need to be in port 3000? HTTPtest is able to start on a random port and ask it which address it's listening on using the .URL property.

Seems so... As we run the region as `Local` for testing that the new relic go client is harcoding the port [here](https://github.com/newrelic/newrelic-client-go/blob/0b2cf7ffb16cc2eba775d937a0fc6be3d73450cd/pkg/region/region_constants.go#L54).

I documented it at the beginning of the integration test, but I'll check if it's better right here in the `runHTTPTestServer`.